dask/highlevelgraph.py

```diff
@@ -923,43 +1141,40 @@ def cull(self, keys: Iterable) -> HighLevelGraph:
         hlg: HighLevelGraph
             Culled high level graph
         """
+        # Set of requested keys
         keys_set = set(flatten(keys))
 
-        all_ext_keys = self.get_all_external_keys()
```

Comment:
@madsbk @gjoseph92 - I decided to experiment a bit with culling in this PR (mostly to make sure the refactor doesn't add any new obstacles to addressing #8570).  I have not done much testing yet, but the most recent changes now avoid collecting external dependencies up front (and instead gather external keys as needed). For a simple `timeseries(...).head()` test with millions of partitions, I definitely see a significant perf bump.

Reply:
>I think this looks like a great approach but I am wondering if it is better to do it in a follow up PR? This PR is getting very large :)

Yes - It is very likely that I will break that part out, but I wanted to make sure the other re-org here is consistent with those changes.